### ISSUE TITLE
Fn name conflict ?

### ISSUE THREAD

**Issue by BenoitBorlee** [rust-analyzer/rust-analyzer#5826]
The following code raises **Expected 1 argument, found 0** in the last `fn`, but it builds fine with cargo.

```
pub trait Cell {
    fn val_str(&self) -> &str;
}
pub trait Row<'a>
where
    Self::CellType: Cell,
{
    type CellType;
    fn val_str(&self, col: usize) -> &str;
    fn get_cell(&'a self, col: usize) -> Self::CellType;
}

pub fn test_row<R: for<'a> Row<'a>>(r: R) {
    let _ = r.get_cell(0).val_str();
}
```

With this fn renaming (`val_str`->`val_str2`) in `Cell` trait , the next code is accepted :

```
pub trait Cell {
    fn val_str2(&self) -> &str;
}
pub trait Row<'a>
where
    Self::CellType: Cell,
{
    type CellType;
    fn val_str(&self, col: usize) -> &str;
    fn get_cell(&'a self, col: usize) -> Self::CellType;
}

pub fn test_row<R: for<'a> Row<'a>>(r: R) {
    let _ = r.get_cell(0).val_str2();
}
```

**Comment by bjorn3** [comment]
What is the type of `r.get_cell(0)` rust-analyzer infered?

**Comment by BenoitBorlee** [comment]
Is `{unknown}` the answer you expect ?

**Comment by bjorn3** [comment]
Yes, I assumed that it was something like that. What happens is likely that it fails to normalize the return value of `get_cell` to `Self::CellType` and instead returns `{unknown}`. It then guesses which method called `val_str` is the right one, and makes the wrong choice.

**Comment by flodiebold** [comment]
So the spurious diagnostic is basically a case of what I describe [here](https://github.com/rust-analyzer/rust-analyzer/issues/5419#issuecomment-660071929); it's interesting though that we fail to normalize the return value. It should be normalized to a placeholder (`Row::CellType<R>`) which should then be known to implement `Cell` via implied bounds.

**Comment by flodiebold** [comment]
@BenoitBorlee are you still having this problem? If so, can you provide a minimal reproduction?

**Comment by lnicola** [comment]
Triage: still happens.

![image](https://user-images.githubusercontent.com/308347/102893400-3026bf80-446a-11eb-9ad4-cb9c89811542.png)

Looks like a problem with the HRTB.

**Comment by lnicola** [comment]
Triage: we now resolve `Self::CellType`, but not `get_cell`:

![image](https://user-images.githubusercontent.com/308347/113902042-271eeb80-97d8-11eb-9ea6-2903700ff4f2.png)

**Comment by Veykril** [comment]
This seems to work now

**Comment by lnicola** [comment]
Awesome, now I wonder what fixed it :smile:.
